src/flagged/applyComplexClasses.js

```diff
@@ -0,0 +1,268 @@
+      rule.each(child => {
+        if (child.type === 'atrule' && child.name === 'apply') {
+          applyRules.unshift(child)
+        }
+      })
+
+      applyRules.forEach(applyRule => {
+        const [
+          importantEntries,
+          applyUtilityNames,
+          important = importantEntries.length > 0,
+        ] = _.partition(applyRule.params.split(' '), n => n === '!important')
+
+        const currentUtilityNames = extractUtilityNames(rule.selector)
+
+        if (_.intersection(applyUtilityNames, currentUtilityNames).length > 0) {
+          const currentUtilityName = _.intersection(applyUtilityNames, currentUtilityNames)[0]
+          throw rule.error(
+            `You cannot \`@apply\` the \`${currentUtilityName}\` utility here because it creates a circular dependency.`
+          )
+        }
+
+        // Extract any post-apply declarations and re-insert them after apply rules
+        const afterRule = rule.clone({ raws: {} })
+        afterRule.nodes = afterRule.nodes.slice(rule.index(applyRule) + 1)
+        rule.nodes = rule.nodes.slice(0, rule.index(applyRule) + 1)
+
+        // Sort applys to match CSS source order
+        const applys = extractUtilityRules(applyUtilityNames, applyRule)
+
+        // Get new rules with the utility portion of the selector replaced with the new selector
+        const rulesToInsert = [
+          ...applys.map(applyUtility => {
+            return generateRulesFromApply(applyUtility, rule.selector)
+          }),
+          afterRule,
+        ]
+
+        const { nodes } = _.tap(postcss.root({ nodes: rulesToInsert }), root =>
+          root.walkDecls(d => (d.important = important))
+        )
+
+        const mergedRules = mergeAdjacentRules(rule, nodes)
+
+        applyRule.remove()
+        rule.after(mergedRules)
+      })
+
+      // If the base rule has nothing in it (all applys were pseudo or responsive variants),
+      // remove the rule fuggit.
+      if (rule.nodes.length === 0) {
+        rule.remove()
+      }
+    })
+  }
+
+  return css
+}
+
+export default function applyComplexClasses(config, getProcessedPlugins) {
+  return function(css) {
+    // Tree already contains @tailwind rules, don't prepend default Tailwind tree
+    if (hasAtRule(css, 'tailwind')) {
+      return processApplyAtRules(css, css, config)
+    }
+
+    // Tree contains no @tailwind rules, so generate all of Tailwind's styles and
+    // prepend them to the user's CSS. Important for <style> blocks in Vue components.
+    return postcss([
+      substituteTailwindAtRules(config, getProcessedPlugins()),
+      evaluateTailwindFunctions(config),
+      substituteVariantsAtRules(config, getProcessedPlugins()),
+      substituteResponsiveAtRules(config),
+      convertLayerAtRulesToControlComments(config),
+      substituteScreenAtRules(config),
+    ])
+      .process(
+        `
+        @tailwind base;
+        @tailwind components;
+        @tailwind utilities;
```

Comment:
So this means custom tailwind declarations won’t be available in VUE components?

Also does this need to be done again and again for each component?

Reply:
What do you mean by "custom tailwind declarations"? If you mean custom CSS classes created in like a `main.css` somewhere then yes, and this has been the case for years and is the fault of the build tooling on the webpack/whatever side.

Tools like Vue run PostCSS _in isolation_ for every single `<style>` block in a Vue component. So Tailwind is "firewalled" inside of that process and has no idea you even _have_ a `main.css` file annoyingly. In a perfect world, the build tooling on the Vue side would concatenate all of the CSS and then run PostCSS _once_ for all of the CSS together. Unfortunately it doesn't so we have to do performance-anti-pattern work-arounds like this to meet users' expectations of things like `@apply` still working.

And unfortunately

Reply:
I mean if you have more `@tailwind foo` declarations than the default 3 (or even less if you Eg don’t want the base styles) [edge cases I know].

I mean in theory you could cheat using the file system or changing the post css configuration on the fly to point to a cache file of the rules.